src/vsr/replica_format.zig

```diff
@@ -27,205 +23,288 @@ pub fn format(
-                );
-                self.formatting = true;
-                while (self.formatting) storage.run();
+            const sector_offset = @divExact(
+                storage_write.zone.offset(write.issued_offset),
+                constants.sector_size,
+            );
+            const sector_count = @divExact(write.issued_buffer.len, constants.sector_size);
+
+            for (sector_offset..sector_offset + sector_count) |sector| {
+                self.sectors_written.set(sector);
             }
-        }
 
-        fn write_sectors_callback(write: *Storage.Write) void {
-            const self: *ReplicaFormat = @alignCast(@fieldParentPtr("write", write));
-            assert(self.formatting);
-            self.formatting = false;
+            if (self.writes_pending == 0) {
+                self.formatting = false;
+            }
         }
 
         fn format_superblock_callback(superblock_context: *SuperBlock.Context) void {
             const self: *ReplicaFormat =
                 @alignCast(@fieldParentPtr("superblock_context", superblock_context));
-            assert(self.formatting);
-            self.formatting = false;
+            assert(self.formatting_superblock);
+            self.formatting_superblock = false;
+        }
+
+        fn verify_writes(self: *ReplicaFormat) void {
+            assert(!self.formatting and !self.formatting_superblock);
+            assert(self.writes_pending == 0);
+
+            assert(self.sectors_written.count() > 0);
+            assert(self.sectors_written.capacity() ==
+                @divExact(data_file_size_min, constants.sector_size));
+
+            // Expect that:
+            // * every sector in the wal_headers zone has been written,
+            // * the first sector in every constants.message_size_max has been written,
+            // * nothing else has been written.
+            //
+            // This might seem to miss the superblock zone, but that's handled entirely by
+            // superblock.zig, which reads back the headers to validate it has been written
+            // correctly.
+            for (0..self.sectors_written.capacity()) |sector| {
+                const sector_start = sector * constants.sector_size;
+
+                if (sector_start >= vsr.Zone.wal_headers.start() and
+                    sector_start < vsr.Zone.wal_headers.start() + vsr.Zone.wal_headers.size().?)
+                {
+                    assert(self.sectors_written.isSet(sector));
+                } else if (sector_start >= vsr.Zone.wal_prepares.start() and
+                    (sector_start - vsr.Zone.wal_prepares.start()) %
+                        constants.message_size_max == 0 and
+                    sector_start < vsr.Zone.wal_prepares.start() + vsr.Zone.wal_prepares.size().?)
+                {
+                    assert(self.sectors_written.isSet(sector));
+                } else {
+                    assert(!self.sectors_written.isSet(sector));
+                }
```

Comment:
These if conditions are somewhat complex, and it isn't obvious that we indeed handle all the cases exhaustively. 

Let maybe do something like this instead:

```
const zone = for (std.enums.values(vsr.Zone)) |zone| {
} else unreachable; 

switch zone {
}
```

to separate out the question of which zone the write belongs two, from the question of whether it should be written or not.